### DISCUSSION TITLE
Primary keys in SQLite should not be autoincrement by default

### DISCUSSION THREAD

**Discussion by fhucho** [mikro-orm/mikro-orm#2657] [Ideas]
Consider not adding `autoincrement` to primary keys when generating SQLite schema. The docs (https://www.sqlite.org/autoinc.html) advise against it:

> 1. The AUTOINCREMENT keyword imposes extra CPU, memory, disk space, and disk I/O overhead and should be avoided if not strictly needed. It is usually not needed.
> 2. In SQLite, a column with type INTEGER PRIMARY KEY is an alias for the ROWID (except in WITHOUT ROWID tables) which is always a 64-bit signed integer.
> 3. On an INSERT, if the ROWID or INTEGER PRIMARY KEY column is not explicitly given a value, then it will be filled automatically with an unused integer, usually one more than the largest ROWID currently in use. This is true regardless of whether or not the AUTOINCREMENT keyword is used.
> 4. If the AUTOINCREMENT keyword appears after INTEGER PRIMARY KEY, that changes the automatic ROWID assignment algorithm to prevent the reuse of ROWIDs over the lifetime of the database. In other words, the purpose of AUTOINCREMENT is to prevent the reuse of ROWIDs from previously deleted rows.

**Comment by B4nan** [comment]
I was hoping this issue will get some traction, as it sounds quite weird to me - it would basically mean allowing the reuse of a (deleted) PKs, right? I don't really use sqlite myself, I guess it makes sense for a locally stored database, but it sounds like a weird default behaviour. 

(also not sure if it's even possible to do with knex, but we could work that around most probably)

Do you know how other similar libraries behave in this manner?

**Comment by fhucho** [comment]
Yes, the ids of previously deleted rows can be reused if the column isn't AUTOINCREMENT. No idea about other libraries...

But this is just a minor issue I randomly noticed, not a big deal. Another cosmetic issue is that the schema uses VARCHAR instead of TEXT. SQLite maps VARCHAR to TEXT for compatibility, but ideally, the native type should be used, it would feel nicer.

**Comment by piotr-cz** [comment]
I found out interesting behavior for sqlite driver (I use to run tests on it):

```ts
@PrimaryKey()
id!: number
// `id` integer not null primary key autoincrement
```

```ts
@PrimaryKey({ type: 'int' })
id!: number
// `id` text not null
```


```ts
@PrimaryKey({ type: 'int', autoincrement: true })
id!: number
// `id` integer not null primary key autoincrement
```

So if one wants to explicitly specify parameter type, then also should add `autoincrement: true`

**Reply by B4nan** [reply]
I think this is just an omission in the default mapping - `type: 'integer'` should work out of box - let me know if that works, will add the `int` there too.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/platforms/Platform.ts#L243

Note that the `type` option is used for many things at once, the original purpose was to provide the TS type, e.g. number/boolean/string, if there is no match found in the mapped types, the value is used as `columnType` - which is another option you can use to just override the column type, and keep the rest working. So `columnType: 'int'` should also work fine, as the `type` will get inferred from decorator metadata.

**Reply by piotr-cz** [reply]
Thanks, Both `type: integer` and `columnType: 'int'` work fine.
I've picked `int` as it's the first option that VSCode Intellisense shows

For future reference: I should have picked on of [Types provided by MikroORM](https://mikro-orm.io/docs/custom-types#types-provided-by-mikroorm)

**Reply by B4nan** [reply]
Yeah the TS types are a mixture of everything supported, `int` should work fine in mysql/mariadb and postgres, those are overriding the default mapping to support it.
